### PR TITLE
Separate system and basis

### PR DESCRIPTION
**Pull request by Schoyen**
In order to support various types of orbital systems (spatial orbitals, spin-orbitals, and general spin-orbitals) we separate the system class from a new `BasisSet`-class. The latter resembles the old `QuantumSystem`, but it does not keep track of the number of occupied basis functions and merely provides methods for the matrix elements and the single-particle functions that apply to all elements. We've also included two new systems called `SpatialOrbitalSystem` and `GeneralOrbitalSystem`. The former is considered a closed-shell system and does not include any explicit spin-dependence. The latter is similar to how the original `QuantumSystem` was intended. Furthermore, this makes the implemented quantum-dots subclasses of `BasisSet`. To create a system you pass in a `BasisSet` and the number of particles (__Note:__ the number of particles is _not_ the same as the number of occupied basis functions in a `SpatialOrbitalSystem`. `SpatialOrbitalSystem` converts the number of particles passed in to the number of occupied basis functions by halving the number of particles) to a system class.

This pull request implements most of these changes along with updated tests, however systems such as the two-dimensional quantum dots need to be updated.

### CODE REVIEW

**Comment by Schoyen** [comment]
This pull request addresses the following issues:

1. It should hopefully close #26 as the only requirement for `l % 2 == 0` is when going from a basis without spin to a `GeneralOrbitalSystem`. This transition makes `l` an even number. However, there are no limitations as to the number of basis functions in the closed shell system and `l` can be odd.
2. Upon a basis change the system will always update the system size thus closing #20. This can be seen here for `QuantumSystem`: https://github.com/Schoyen/quantum-systems/blob/869cc61811183d2f2be986cecf3882bfd21c7c5d/quantum_systems/system.py#L66-L68 and here for `BasisSet`: https://github.com/Schoyen/quantum-systems/blob/869cc61811183d2f2be986cecf3882bfd21c7c5d/quantum_systems/basis_set.py#L270-L314
3. Much of the functionality in the `system_helper.py`-program has been moved into `BasisSet` and `QuantumSystem` as static methods as discussed in #28.

File: quantum_systems/quantum_dots/two_dim/two_dim_ho.py [hunk 1]
```diff
@@ -80,30 +74,18 @@ def __init__(
         self.radius = np.linspace(0, self.radius_length, self.num_grid_points)
         self.theta = np.linspace(0, 2 * np.pi, self.num_grid_points)
```
Comment (halvarsu) [inline quantum_systems/quantum_dots/two_dim/two_dim_ho.py]: Is there a reason not to call `setup_basis()` here, as is done in the 1D case? Makes usage of this basis set unintuitively different.

File: quantum_systems/spatial_orbital_system.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+    where :math:`x = (\mathbf{r}, m_s)` is a generalized coordinate of position
+    and spin, and :math:`sigma(m_s)` is either :math:`\alpha(m_s)` or
+    :math:`\beta(m_s)` as the two-dimensional spin basis states. This means
+    that we only store :math:`psi(mathbf{r}, t)`.
+
+    Parameters
+    ----------
+    n : int
+        Number of occupied basis functions. Note that this should correspond to
```
Comment (halvarsu) [inline quantum_systems/spatial_orbital_system.py]: A bit unclear, as number of particles `!=` number of occupied basis functions. I think it should be called number of particles, maybe by another name (`N`?) and then changed to `n = N // 2` internally. 

Alternatively one can send in the number of basis functions `n`, and then never talk about particles.

File: quantum_systems/quantum_dots/two_dim/two_dim_ho.py [hunk 3]
```diff
@@ -80,30 +74,18 @@ def __init__(
...
-        anti_symmetrize=True,
-        verbose=True,
-        cast_to_complex=True,
-    ):
-        self._h = self.omega * get_one_body_elements(self.l // 2)
+    def setup_basis(self,):
+        self._h = self.omega * get_one_body_elements(self.l)
         self._u = np.sqrt(self.omega) * get_coulomb_elements(
-            self.l // 2, verbose=verbose
+            self.l, verbose=verbose
```
Comment (halvarsu) [inline quantum_systems/quantum_dots/two_dim/two_dim_ho.py]: verbose is not defined

**Comment by halvarsu** [comment]
Great work!

File: quantum_systems/spatial_orbital_system.py (hunk 2 above)
Comment (Schoyen) [reply quantum_systems/spatial_orbital_system.py]: Hmm, agreed. In practice `n` is interpreted as the number of particles upon initialization, but internally changed to `n // 2`. Perhaps just stating that `n` is the number of particles in the documentation is enough? We can also add a doctest demonstrating this, viz.
```python
>>> n = 4
>>> l = 20
>>> dim = 2
>>> spas = SpatialOrbitalSystem(n, RandomBasisSet(l, dim))
>>> spas.n == n // 2
True
```

File: quantum_systems/quantum_dots/two_dim/two_dim_ho.py (hunk 1 above)
Comment (Schoyen) [reply quantum_systems/quantum_dots/two_dim/two_dim_ho.py]: The reason is laziness :P

File: quantum_systems/quantum_dots/two_dim/two_dim_ho.py (hunk 3 above)
Comment (Schoyen) [reply quantum_systems/quantum_dots/two_dim/two_dim_ho.py]: Same as the other two-dim trouble.

**Pull request merged**